### PR TITLE
default tab count to 1 with home tab

### PR DESCRIPTION
**Pull request by brindy**
Task/Issue URL: https://app.asana.com/0/392891325557410/793443334130013
Tech Design URL:
CC:

**Description**:

Adds the concept of a "home tab", which is a tab with no link of which there can be only one.

( Note: includes #568 to try to avoid merge conflicts )

**Steps to test this PR**:

Broad testing required around:

- Adding and removing tabs
- Navigating and opening links in a new tab
- Navigating and opening links in the background
- Opening the about page (Settings -> About -> tap link)
- Opening links from outside the app (e.g. Safari -> Share -> DDG
- Opening bookmarks
- All of the above while there's a "home tab" open
- Migration (from previous version of app to this version)

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

**Review by bwaresiak — COMMENTED** [review]
This count change makes much sense. :) Still playing with it, but I've found some problems:

1. Adding multiple Home Screen tabs - I believe we should be able to close these in tab switcher.
2. Bug: add multiple home screen tabs, then enter url -> it won't appear on current tab but somewhere "underneath".
3. When testing #2, I've managed to crash it:

```Fatal error: Duplicate keys of type 'Tab' were found in a Dictionary.
This usually means either that the type violates Hashable's requirements, or
that members of such a dictionary were mutated after insertion.
2020-02-10 14:40:30.668141+0100 DuckDuckGo[15130:978817] Fatal error: Duplicate keys of type 'Tab' were found in a Dictionary.```

**Comment by brindy** [comment]
Thanks! That's my fault for trying to make a quick change before running off to do an errand.  Will fix.

File: DuckDuckGo/MainViewController.swift [hunk 1]
```diff
@@ -346,11 +346,14 @@ class MainViewController: UIViewController {
     }
 
     func loadUrlInNewTab(_ url: URL) {
+        print(#function, "***")
```
Comment (tomasstrba) [inline DuckDuckGo/MainViewController.swift]: is this output for debug purposes?(maybe change to os_log) or did you forget to remove it?

File: DuckDuckGo/TabManager.swift [hunk 2]
```diff
@@ -105,10 +121,11 @@ class TabManager {
         let controller = buildController(forTab: tab, url: url)
         tabControllerCache.append(controller)
 
-        if let index = model.currentIndex, inBackground {
+        let index = model.currentIndex
+        if inBackground {
             model.insert(tab: tab, at: index + 1)
         } else {
-            model.add(tab: tab)
+            model.insert(tab: tab, at: model.count - 1)
```
Comment (tomasstrba) [inline DuckDuckGo/TabManager.swift]: model.last computed property would be more readable and safer to have here. But not necessary :)

File: DuckDuckGo/TabViewController.swift [hunk 3]
```diff
@@ -263,21 +276,24 @@ class TabViewController: UIViewController {
         
         switch keyPath {
             
-        case WebViewKeyPaths.estimatedProgress:
+        case #keyPath(WKWebView.estimatedProgress):
```
Comment (tomasstrba) [inline DuckDuckGo/TabViewController.swift]: nice :)

**Review by tomasstrba — APPROVED** [review]
LGTM! Good job 👍  (I found just one forgotten print mentioned in comment)

**Comment by brindy** [comment]
Thanks @tomasstrba! @bwaresiak also made a suggestion so I've addressed that, your comments and Bartek will re-test it.  Thanks!

**Comment by bwaresiak** [comment]
Glad to see opening new tabs from link (both foreground and background ones) works in unified way now :) 

Unfortunately, there is one problem I've noticed:
1. Open a website.
2. Open a link in background tab.
3. Navigate to tab switcher.

Expected: Current tab is selected, background tab should be marked with 'dot'.
Actual: Current tab is not selected, background tab is marked with 'dot' and selected.

**Review by bwaresiak — APPROVED** [review]
LGTM!

**Comment by brindy** [comment]
Thanks @bwaresiak and @tomasstrba

**Pull request merged**